## 1. Data Preparation

## i. Load The Dataset

In [1]:
from warnings import filterwarnings
filterwarnings('ignore')

In [2]:
import pandas as pd

In [3]:
df1 = pd.read_csv('Volunteer_Activities.csv',na_values=(' '),encoding='latin-1')

In [4]:
df1.head()

Title             Event Type  \
0                          Become PCA's Chapter Head      Community Service   
1                          Blood Donation Drive 2024  Social Responsibility   
2                                  Meals That Matter  Social Responsibility   
3       Volunteer Flagship Training Program Batch 39      Community Service   
4  Handwritten Assignment Support for Eye-Impaire...  Social Responsibility   

                        Date                Time     City  \
0  Oct 28 2024 - Oct 28 2025   9:00 AM - 5:00 PM   Lahore   
1  Jan 09 2024 - Dec 31 2024   9:00 AM - 5:00 PM   Lahore   
2  Aug 16 2024 - Dec 01 2024  7:00 AM - 11:59 AM   Lahore   
3  Oct 28 2024 - Nov 05 2024   3:00 PM - 5:00 AM   Lahore   
4  Oct 15 2024 - Oct 31 2024  2:00 AM - 12:00 PM  Sahiwal   

                 Organization  Rating  Volunteers Engaged  Min Duration  \
0  Pakistan Citizens Alliance       3                  17             6   
1           Sundas foundation       5                  12             2   
2         Meherban Foundation       5                   2             9   
3  Pakistan Citizens Alliance       5                  19             5   
4       University of Sahiwal       3                   4             9   

   Max Duration Duration String  
0            18    6 - 18 hours  
1            18    2 - 18 hours  
2            20    9 - 20 hours  
3            16    5 - 16 hours  
4            17    9 - 17 hours

In [5]:
df1.shape

(702, 11)

In [6]:
df2 = pd.read_csv('Volunteers_Data.csv',na_values=(' '),encoding='latin-1')

In [7]:
df2.head()

User ID            Name  Gender  Age    City Province  Rating  \
0    10297    Farhana Bano  Female   24  Lahore   Punjab     4.7   
1    12193  khalid Hussain    Male   21  Sukkur    Sindh     4.0   
2    10583        Ali Awan    Male   23  Lahore   Punjab     5.0   
3    10248          Barira  Female   23  Lahore   Punjab     4.6   
4    12052  Muhammad Ahmed    Male   20  Lahore   Punjab     5.0   

                                     Degree  \
0                  Bachelors/ BS (16 years)   
1                      Bachelors (14 years)   
2                      Bachelors (14 years)   
3                  Bachelors/ BS (16 years)   
4  Higher Secondary (Intermediate/A levels)   

                                        Field  \
0                                 Engineering   
1  Computer Sciences & Information Technology   
2  Computer Sciences & Information Technology   
3                                 Engineering   
4  Computer Sciences & Information Technology   

                               University Employment Status  \
0  University of Engineering & Technology        Unemployed   
1                                   Other           Student   
2   University of Management & Technology           Student   
3  University of Engineering & Technology           Student   
4                                   Other           Student   

                                                 Bio  \
0  My name is farhana bano. I am doing chemical\r...   
1  i am student of bsit at Akhuwat college univer...   
2                 Looking forward for some work !!!!   
3  I am a very passionate person. I will do my wo...   
4  As well as studies I show  my interest in  spe...   

                                              Skills  \
0  Arts and Crafts, Event Management, Presentatio...   
1                                        Fundraising   
2  Leadership, Mentoring, Photography, Team Manag...   
3  Branding & Marketing, Event Management, Organi...   
4                                         Leadership   

                                Causes Interested In  \
0  Charity, Education, Poverty Eradication, Women...   
1                                     Social Service   
2                                Poverty Eradication   
3  Climate Change, Education, Social Service, Stu...   
4                                Poverty Eradication   

                                         Event Types  Activities Volunteered  \
0  Distribution of Goods or Ration, Events, Confe...                       8   
1                          Fundraising and Donations                       1   
2                              Social Responsibility                       8   
3  Community Service, Distribution of Goods or Ra...                       8   
4                                Events, Conferences                       1   

   Reviews Received  Referral Count  Hours Completed  
0                 7               0               24  
1                 1               0               40  
2                 6              98               38  
3                 7               0               23  
4                 1               0               40

In [8]:
df2.shape

(7686, 19)

In [9]:
df1.describe().T

count       mean       std   min   25%   50%   75%   max
Rating              702.0   3.854701  1.036550   1.0   3.0   4.0   5.0   5.0
Volunteers Engaged  702.0   9.424501  5.760991   1.0   5.0   8.0  15.0  20.0
Min Duration        702.0   5.619658  2.842878   1.0   3.0   6.0   8.0  10.0
Max Duration        702.0  15.559829  2.866748  11.0  13.0  16.0  18.0  20.0

In [10]:
df2.describe().T

count          mean          std    min      25%  \
User ID                 7686.0  10761.323185  5946.387164  102.0  5179.75   
Age                     7686.0     26.514442    13.773282   10.0    22.00   
Rating                  7686.0      0.622756     1.555353    0.0     0.00   
Activities Volunteered  7686.0      0.672782     2.018717    0.0     0.00   
Reviews Received        7686.0      0.321233     1.190609    0.0     0.00   
Referral Count          7686.0      0.360265     4.491841    0.0     0.00   
Hours Completed         7686.0      1.927270     9.190886    0.0     0.00   

                            50%       75%      max  
User ID                 10752.0  15348.75  21259.0  
Age                        24.0     28.00   1009.0  
Rating                      0.0      0.00      5.0  
Activities Volunteered      0.0      1.00     59.0  
Reviews Received            0.0      0.00     20.0  
Referral Count              0.0      0.00    161.0  
Hours Completed             0.0      0.00    212.0

## ii. Data Cleaning

In [11]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Title               702 non-null    object
 1   Event Type          702 non-null    object
 2   Date                702 non-null    object
 3   Time                702 non-null    object
 4   City                421 non-null    object
 5   Organization        702 non-null    object
 6   Rating              702 non-null    int64 
 7   Volunteers Engaged  702 non-null    int64 
 8   Min Duration        702 non-null    int64 
 9   Max Duration        702 non-null    int64 
 10  Duration String     702 non-null    object
dtypes: int64(4), object(7)
memory usage: 60.5+ KB


In [12]:
df1.isnull().sum()

Title                   0
Event Type              0
Date                    0
Time                    0
City                  281
Organization            0
Rating                  0
Volunteers Engaged      0
Min Duration            0
Max Duration            0
Duration String         0
dtype: int64

In [13]:
df1.duplicated().sum()

0

In [14]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7686 entries, 0 to 7685
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   User ID                 7686 non-null   int64  
 1   Name                    7686 non-null   object 
 2   Gender                  7686 non-null   object 
 3   Age                     7686 non-null   int64  
 4   City                    7686 non-null   object 
 5   Province                7686 non-null   object 
 6   Rating                  7686 non-null   float64
 7   Degree                  7686 non-null   object 
 8   Field                   7686 non-null   object 
 9   University              7683 non-null   object 
 10  Employment Status       7686 non-null   object 
 11  Bio                     7682 non-null   object 
 12  Skills                  7677 non-null   object 
 13  Causes Interested In    7686 non-null   object 
 14  Event Types             7686 non-null   

In [15]:
df2.isnull().sum()

User ID                   0
Name                      0
Gender                    0
Age                       0
City                      0
Province                  0
Rating                    0
Degree                    0
Field                     0
University                3
Employment Status         0
Bio                       4
Skills                    9
Causes Interested In      0
Event Types               0
Activities Volunteered    0
Reviews Received          0
Referral Count            0
Hours Completed           0
dtype: int64

In [16]:
df2.duplicated().sum()

1

So, After performing some operations, we got to know that there are missing and duplicate values in both datasets.

## Handling Missing Values

In **Volunteer_Activities.csv**, there are 281 missing values for **Column = 'City'**, so filling it with default value of **City = 'Karachi'**.

In [17]:
df1['City'] = df1['City'].fillna('Karachi')

In [18]:
df1.isnull().sum()

Title                 0
Event Type            0
Date                  0
Time                  0
City                  0
Organization          0
Rating                0
Volunteers Engaged    0
Min Duration          0
Max Duration          0
Duration String       0
dtype: int64

In [19]:
df1.shape

(702, 11)

In **Volunteers_Data.csv**, there are missing values for **Column = 'University'**, **'Bio'**, **'Skills'**, so dropping these rows with missing values as the number of rows are not too high.

In [20]:
df2 = df2.dropna(subset=['University', 'Bio', 'Skills'])

In [21]:
df2.isnull().sum()

User ID                   0
Name                      0
Gender                    0
Age                       0
City                      0
Province                  0
Rating                    0
Degree                    0
Field                     0
University                0
Employment Status         0
Bio                       0
Skills                    0
Causes Interested In      0
Event Types               0
Activities Volunteered    0
Reviews Received          0
Referral Count            0
Hours Completed           0
dtype: int64

In [22]:
df2.shape

(7670, 19)

## Handling Duplicate Rows

As in both datasets, there is only one duplicate row in **Volunteers_Data.csv**, so dropping it.

In [23]:
df2.drop_duplicates(inplace=True)

In [24]:
df2.duplicated().sum()

0

In [25]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Title               702 non-null    object
 1   Event Type          702 non-null    object
 2   Date                702 non-null    object
 3   Time                702 non-null    object
 4   City                702 non-null    object
 5   Organization        702 non-null    object
 6   Rating              702 non-null    int64 
 7   Volunteers Engaged  702 non-null    int64 
 8   Min Duration        702 non-null    int64 
 9   Max Duration        702 non-null    int64 
 10  Duration String     702 non-null    object
dtypes: int64(4), object(7)
memory usage: 60.5+ KB


In [26]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7669 entries, 0 to 7685
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   User ID                 7669 non-null   int64  
 1   Name                    7669 non-null   object 
 2   Gender                  7669 non-null   object 
 3   Age                     7669 non-null   int64  
 4   City                    7669 non-null   object 
 5   Province                7669 non-null   object 
 6   Rating                  7669 non-null   float64
 7   Degree                  7669 non-null   object 
 8   Field                   7669 non-null   object 
 9   University              7669 non-null   object 
 10  Employment Status       7669 non-null   object 
 11  Bio                     7669 non-null   object 
 12  Skills                  7669 non-null   object 
 13  Causes Interested In    7669 non-null   object 
 14  Event Types             7669 non-null   objec

## iii. Converting Data Types

In [27]:
df1['Organization'] = df1['Organization'].astype('category')
df1['Rating'] = df1['Rating'].astype('float64')

In [28]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   Title               702 non-null    object  
 1   Event Type          702 non-null    object  
 2   Date                702 non-null    object  
 3   Time                702 non-null    object  
 4   City                702 non-null    object  
 5   Organization        702 non-null    category
 6   Rating              702 non-null    float64 
 7   Volunteers Engaged  702 non-null    int64   
 8   Min Duration        702 non-null    int64   
 9   Max Duration        702 non-null    int64   
 10  Duration String     702 non-null    object  
dtypes: category(1), float64(1), int64(3), object(6)
memory usage: 61.9+ KB


In [29]:
df2['Gender'] = df2['Gender'].astype('category')
df2['Province'] = df2['Province'].astype('category')
df2['Employment Status'] = df2['Employment Status'].astype('category')

In [30]:
df2['Degree'] = df2['Degree'].astype('category')
df2['Field'] = df2['Field'].astype('category')
df2['University'] = df2['University'].astype('category')

In [31]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7669 entries, 0 to 7685
Data columns (total 19 columns):
 #   Column                  Non-Null Count  Dtype   
---  ------                  --------------  -----   
 0   User ID                 7669 non-null   int64   
 1   Name                    7669 non-null   object  
 2   Gender                  7669 non-null   category
 3   Age                     7669 non-null   int64   
 4   City                    7669 non-null   object  
 5   Province                7669 non-null   category
 6   Rating                  7669 non-null   float64 
 7   Degree                  7669 non-null   category
 8   Field                   7669 non-null   category
 9   University              7669 non-null   category
 10  Employment Status       7669 non-null   category
 11  Bio                     7669 non-null   object  
 12  Skills                  7669 non-null   object  
 13  Causes Interested In    7669 non-null   object  
 14  Event Types             7669 

## iv. Feature Engineering

## Encoding Categorical Features

In [32]:
df1['Organization'] = df1['Organization'].cat.codes

In [33]:
df1.head()

Title             Event Type  \
0                          Become PCA's Chapter Head      Community Service   
1                          Blood Donation Drive 2024  Social Responsibility   
2                                  Meals That Matter  Social Responsibility   
3       Volunteer Flagship Training Program Batch 39      Community Service   
4  Handwritten Assignment Support for Eye-Impaire...  Social Responsibility   

                        Date                Time     City  Organization  \
0  Oct 28 2024 - Oct 28 2025   9:00 AM - 5:00 PM   Lahore           112   
1  Jan 09 2024 - Dec 31 2024   9:00 AM - 5:00 PM   Lahore           153   
2  Aug 16 2024 - Dec 01 2024  7:00 AM - 11:59 AM   Lahore            88   
3  Oct 28 2024 - Nov 05 2024   3:00 PM - 5:00 AM   Lahore           112   
4  Oct 15 2024 - Oct 31 2024  2:00 AM - 12:00 PM  Sahiwal           179   

   Rating  Volunteers Engaged  Min Duration  Max Duration Duration String  
0     3.0                  17             6            18    6 - 18 hours  
1     5.0                  12             2            18    2 - 18 hours  
2     5.0                   2             9            20    9 - 20 hours  
3     5.0                  19             5            16    5 - 16 hours  
4     3.0                   4             9            17    9 - 17 hours

In [34]:
df2['Gender'] = df2['Gender'].cat.codes
df2['Province'] = df2['Province'].cat.codes
df2['Employment Status'] = df2['Employment Status'].cat.codes

In [35]:
df2['Degree'] = df2['Degree'].cat.codes
df2['Field'] = df2['Field'].cat.codes
df2['University'] = df2['University'].cat.codes

In [36]:
df2.head()

User ID            Name  Gender  Age    City  Province  Rating  Degree  \
0    10297    Farhana Bano       0   24  Lahore         5     4.7       4   
1    12193  khalid Hussain       1   21  Sukkur         6     4.0       2   
2    10583        Ali Awan       1   23  Lahore         5     5.0       2   
3    10248          Barira       0   23  Lahore         5     4.6       4   
4    12052  Muhammad Ahmed       1   20  Lahore         5     5.0       7   

   Field  University  Employment Status  \
0      3         437                  5   
1      2         287                  4   
2      2         458                  4   
3      3         437                  4   
4      2         287                  4   

                                                 Bio  \
0  My name is farhana bano. I am doing chemical\r...   
1  i am student of bsit at Akhuwat college univer...   
2                 Looking forward for some work !!!!   
3  I am a very passionate person. I will do my wo...   
4  As well as studies I show  my interest in  spe...   

                                              Skills  \
0  Arts and Crafts, Event Management, Presentatio...   
1                                        Fundraising   
2  Leadership, Mentoring, Photography, Team Manag...   
3  Branding & Marketing, Event Management, Organi...   
4                                         Leadership   

                                Causes Interested In  \
0  Charity, Education, Poverty Eradication, Women...   
1                                     Social Service   
2                                Poverty Eradication   
3  Climate Change, Education, Social Service, Stu...   
4                                Poverty Eradication   

                                         Event Types  Activities Volunteered  \
0  Distribution of Goods or Ration, Events, Confe...                       8   
1                          Fundraising and Donations                       1   
2                              Social Responsibility                       8   
3  Community Service, Distribution of Goods or Ra...                       8   
4                                Events, Conferences                       1   

   Reviews Received  Referral Count  Hours Completed  
0                 7               0               24  
1                 1               0               40  
2                 6              98               38  
3                 7               0               23  
4                 1               0               40

## Encode Cities Consistently Across Both Datasets

In [37]:
from sklearn.preprocessing import LabelEncoder

In [38]:
# Combine unique city names and fit the encoder
label_encoder = LabelEncoder()
label_encoder.fit(pd.concat([df1['City'], df2['City']]))

LabelEncoder()

In [39]:
# Apply the encoder to both datasets
df1['City'] = label_encoder.transform(df1['City'])
df2['City'] = label_encoder.transform(df2['City'])

In [40]:
# Verify the encoding
print("Encoded Cities in df1:", df1['City'].unique())
print("Encoded Cities in df2:", df2['City'].unique())

Encoded Cities in df1: [ 69 107  59  52  62  83 102 105  78  44  63  97  74  22  47  75  93 101
  50  12  54  71  61 109 103  37  25  31  85  23  92  42 114  29  82  89
 113  66  35  48  14  77  51  65  73 112  17]
Encoded Cities in df2: [ 69 116  92  63 103   0  97 117  78 123  37 108 114  57 105  45  91  42
 112  14  83  47  73  86  23  87 125  12  81  13 101  75  93  59  55 122
  44   9  79 127  18 110 107  51  52  15  50 102  82 121  17  48  25 129
  62  53  90   4  71 124  98  94  89  66  20  49  19  74  85  22  68  26
  54   7  88  10  72  65  28  77  67   2  61 106  43  40   5  56  46  34
  76 115 104 118 119  31 109  36 113 126  60  38  39  11 128 100  35  95
  27  80  70  29  96  64  30  33 120  58  24  32   8  84  21  41  16   3
  99   1 111   6]


In [41]:
df1.head()

Title             Event Type  \
0                          Become PCA's Chapter Head      Community Service   
1                          Blood Donation Drive 2024  Social Responsibility   
2                                  Meals That Matter  Social Responsibility   
3       Volunteer Flagship Training Program Batch 39      Community Service   
4  Handwritten Assignment Support for Eye-Impaire...  Social Responsibility   

                        Date                Time  City  Organization  Rating  \
0  Oct 28 2024 - Oct 28 2025   9:00 AM - 5:00 PM    69           112     3.0   
1  Jan 09 2024 - Dec 31 2024   9:00 AM - 5:00 PM    69           153     5.0   
2  Aug 16 2024 - Dec 01 2024  7:00 AM - 11:59 AM    69            88     5.0   
3  Oct 28 2024 - Nov 05 2024   3:00 PM - 5:00 AM    69           112     5.0   
4  Oct 15 2024 - Oct 31 2024  2:00 AM - 12:00 PM   107           179     3.0   

   Volunteers Engaged  Min Duration  Max Duration Duration String  
0                  17             6            18    6 - 18 hours  
1                  12             2            18    2 - 18 hours  
2                   2             9            20    9 - 20 hours  
3                  19             5            16    5 - 16 hours  
4                   4             9            17    9 - 17 hours

In [42]:
df2.head()

User ID            Name  Gender  Age  City  Province  Rating  Degree  \
0    10297    Farhana Bano       0   24    69         5     4.7       4   
1    12193  khalid Hussain       1   21   116         6     4.0       2   
2    10583        Ali Awan       1   23    69         5     5.0       2   
3    10248          Barira       0   23    69         5     4.6       4   
4    12052  Muhammad Ahmed       1   20    69         5     5.0       7   

   Field  University  Employment Status  \
0      3         437                  5   
1      2         287                  4   
2      2         458                  4   
3      3         437                  4   
4      2         287                  4   

                                                 Bio  \
0  My name is farhana bano. I am doing chemical\r...   
1  i am student of bsit at Akhuwat college univer...   
2                 Looking forward for some work !!!!   
3  I am a very passionate person. I will do my wo...   
4  As well as studies I show  my interest in  spe...   

                                              Skills  \
0  Arts and Crafts, Event Management, Presentatio...   
1                                        Fundraising   
2  Leadership, Mentoring, Photography, Team Manag...   
3  Branding & Marketing, Event Management, Organi...   
4                                         Leadership   

                                Causes Interested In  \
0  Charity, Education, Poverty Eradication, Women...   
1                                     Social Service   
2                                Poverty Eradication   
3  Climate Change, Education, Social Service, Stu...   
4                                Poverty Eradication   

                                         Event Types  Activities Volunteered  \
0  Distribution of Goods or Ration, Events, Confe...                       8   
1                          Fundraising and Donations                       1   
2                              Social Responsibility                       8   
3  Community Service, Distribution of Goods or Ra...                       8   
4                                Events, Conferences                       1   

   Reviews Received  Referral Count  Hours Completed  
0                 7               0               24  
1                 1               0               40  
2                 6              98               38  
3                 7               0               23  
4                 1               0               40

## Encode Event Type in Volunteer Activities:

In [43]:
# Split combined event types in df1
df1['Event Type'] = df1['Event Type'].str.split(', ')

In [44]:
from sklearn.preprocessing import MultiLabelBinarizer

In [45]:
mlb = MultiLabelBinarizer()

event_type_encoded = pd.DataFrame(mlb.fit_transform(df1['Event Type']),
                                  columns=mlb.classes_,
                                  index=df1.index)

# Add the encoded columns back to df1_encoded
df1_encoded = pd.concat([df1, event_type_encoded], axis=1).drop(columns=['Event Type'])

In [46]:
from sklearn.preprocessing import StandardScaler

In [47]:
# Normalize Rating
scaler = StandardScaler()
df1_encoded['Rating'] = scaler.fit_transform(df1_encoded[['Rating']])

In [48]:
print(df1_encoded.columns)

Index(['Title', 'Date', 'Time', 'City', 'Organization', 'Rating',
       'Volunteers Engaged', 'Min Duration', 'Max Duration', 'Duration String',
       'Community Service', 'Conferences', 'Events', 'Social Responsibility',
       'Teaching', 'Workshops'],
      dtype='object')


In [49]:
# Select only the key features
df1_encoded = df1_encoded[['Community Service', 'Conferences', 'Events', 'Social Responsibility',
       'Teaching', 'Workshops', 'City', 'Organization', 'Rating']]

In [50]:
df1_encoded.head()

Community Service  Conferences  Events  Social Responsibility  Teaching  \
0                  1            0       0                      0         0   
1                  0            0       0                      1         0   
2                  0            0       0                      1         0   
3                  1            0       0                      0         0   
4                  0            0       0                      1         0   

   Workshops  City  Organization    Rating  
0          0    69           112 -0.825151  
1          0    69           153  1.105702  
2          0    69            88  1.105702  
3          0    69           112  1.105702  
4          0   107           179 -0.825151

## Encode Event Types in Volunteers Data:

In [51]:
# Multi-valued features to encode
mlb = MultiLabelBinarizer()

# Encode Skills, Causes Interested In, and Event Types
skills_encoded = pd.DataFrame(mlb.fit_transform(df2['Skills'].str.split(', ')),
                               columns=mlb.classes_, index=df2.index)
causes_encoded = pd.DataFrame(mlb.fit_transform(df2['Causes Interested In'].str.split(', ')),
                               columns=mlb.classes_, index=df2.index)
event_types_encoded = pd.DataFrame(mlb.fit_transform(df2['Event Types'].str.split(', ')),
                                    columns=mlb.classes_, index=df2.index)

df2_encoded = pd.concat([df2, skills_encoded, causes_encoded, event_types_encoded], axis=1).drop(columns=['Skills', 'Causes Interested In', 'Event Types'])

In [52]:
df2_encoded.head()

User ID            Name  Gender  Age  City  Province  Rating  Degree  \
0    10297    Farhana Bano       0   24    69         5     4.7       4   
1    12193  khalid Hussain       1   21   116         6     4.0       2   
2    10583        Ali Awan       1   23    69         5     5.0       2   
3    10248          Barira       0   23    69         5     4.6       4   
4    12052  Muhammad Ahmed       1   20    69         5     5.0       7   

   Field  University  ...  Events Fundraising and Donations  March  \
0      3         437  ...       1                         1      0   
1      2         287  ...       0                         1      0   
2      2         458  ...       0                         0      0   
3      3         437  ...       1                         0      0   
4      2         287  ...       1                         0      0   

   Milkar SDG's Champions  Rallies  Social Responsibility  Teaching  Vigils  \
0                       0        0                      0         1       0   
1                       0        0                      0         0       0   
2                       0        0                      1         0       0   
3                       0        0                      0         1       0   
4                       0        0                      0         0       0   

   Walks  Workshops  
0      0          1  
1      0          0  
2      0          0  
3      0          1  
4      0          0  

[5 rows x 103 columns]

In [53]:
print(df2_encoded.columns)

Index(['User ID', 'Name', 'Gender', 'Age', 'City', 'Province', 'Rating',
       'Degree', 'Field', 'University',
       ...
       'Events', 'Fundraising and Donations', 'March',
       'Milkar SDG's Champions', 'Rallies', 'Social Responsibility',
       'Teaching', 'Vigils', 'Walks', 'Workshops'],
      dtype='object', length=103)


In [54]:
# Normalize Rating
df2_encoded['Rating'] = scaler.fit_transform(df2_encoded[['Rating']])

# Combine encoded features with key columns
df2_encoded = pd.concat([df2_encoded[['City', 'Rating']],
                         skills_encoded, causes_encoded, event_types_encoded], axis=1)

In [55]:
df2_encoded.head()

City    Rating  Arts and Crafts  Auditing  Beautician  \
0    69  2.618342                1         0           0   
1   116  2.168675                0         0           0   
2    69  2.811057                0         0           0   
3    69  2.554104                0         0           0   
4    69  2.811057                0         0           0   

   Branding & Marketing  Coaching  Community Outreach  Conflict Management  \
0                     0         0                   0                    0   
1                     0         0                   0                    0   
2                     0         0                   0                    0   
3                     1         0                   0                    0   
4                     0         0                   0                    0   

   Content Writing  ...  Events  Fundraising and Donations  March  \
0                0  ...       1                          1      0   
1                0  ...       0                          1      0   
2                0  ...       0                          0      0   
3                0  ...       1                          0      0   
4                0  ...       1                          0      0   

   Milkar SDG's Champions  Rallies  Social Responsibility  Teaching  Vigils  \
0                       0        0                      0         1       0   
1                       0        0                      0         0       0   
2                       0        0                      1         0       0   
3                       0        0                      0         1       0   
4                       0        0                      0         0       0   

   Walks  Workshops  
0      0          1  
1      0          0  
2      0          0  
3      0          1  
4      0          0  

[5 rows x 89 columns]

In [56]:
print(df2_encoded.columns)

Index(['City', 'Rating', 'Arts and Crafts', 'Auditing', 'Beautician',
       'Branding & Marketing', 'Coaching', 'Community Outreach',
       'Conflict Management', 'Content Writing', 'Cooking/ Baking',
       'Counseling', 'Dancing', 'Data Analysis', 'Data Gathering',
       'Database Management', 'Digital Media Management',
       'Editing/proofreading', 'Event Management', 'Financial Skills',
       'Fundraising', 'Goal Setting', 'Graphic Designing',
       'Health/Medical Experience', 'Human Resource Skills',
       'Innovation & Creativity (New Ideas)', 'Interviewing', 'Leadership',
       'Logistics & Inventory Management', 'Management', 'Mentoring',
       'Negotiations & Efficiency', 'Organizing Events',
       'Pharmaceutical Activities', 'Photography', 'Planning & Execution',
       'Policy Making', 'Presentation Skills', 'Problem Solving',
       'Public Relations', 'Public Speaking', 'Sales',
       'Sign Language Expertise', 'Social Media Strategy & Marketing',
       'Sta

## Columns Alignment

In [57]:
# Add missing columns to df2_encoded
missing_cols_in_df2 = set(df1_encoded.columns) - set(df2_encoded.columns)
for col in missing_cols_in_df2:
    df2_encoded[col] = 0

# Add missing columns to df1_encoded
missing_cols_in_df1 = set(df2_encoded.columns) - set(df1_encoded.columns)
for col in missing_cols_in_df1:
    df1_encoded[col] = 0

# Ensure the column order is the same
df1_encoded = df1_encoded[df2_encoded.columns]

In [58]:
df1_encoded.head()

City    Rating  Arts and Crafts  Auditing  Beautician  \
0    69 -0.825151                0         0           0   
1    69  1.105702                0         0           0   
2    69  1.105702                0         0           0   
3    69  1.105702                0         0           0   
4   107 -0.825151                0         0           0   

   Branding & Marketing  Coaching  Community Outreach  Conflict Management  \
0                     0         0                   0                    0   
1                     0         0                   0                    0   
2                     0         0                   0                    0   
3                     0         0                   0                    0   
4                     0         0                   0                    0   

   Content Writing  ...  Fundraising and Donations  March  \
0                0  ...                          0      0   
1                0  ...                          0      0   
2                0  ...                          0      0   
3                0  ...                          0      0   
4                0  ...                          0      0   

   Milkar SDG's Champions  Rallies  Social Responsibility  Teaching  Vigils  \
0                       0        0                      0         0       0   
1                       0        0                      1         0       0   
2                       0        0                      1         0       0   
3                       0        0                      0         0       0   
4                       0        0                      1         0       0   

   Walks  Workshops  Organization  
0      0          0           112  
1      0          0           153  
2      0          0            88  
3      0          0           112  
4      0          0           179  

[5 rows x 90 columns]

In [59]:
print(df1_encoded.columns)

Index(['City', 'Rating', 'Arts and Crafts', 'Auditing', 'Beautician',
       'Branding & Marketing', 'Coaching', 'Community Outreach',
       'Conflict Management', 'Content Writing', 'Cooking/ Baking',
       'Counseling', 'Dancing', 'Data Analysis', 'Data Gathering',
       'Database Management', 'Digital Media Management',
       'Editing/proofreading', 'Event Management', 'Financial Skills',
       'Fundraising', 'Goal Setting', 'Graphic Designing',
       'Health/Medical Experience', 'Human Resource Skills',
       'Innovation & Creativity (New Ideas)', 'Interviewing', 'Leadership',
       'Logistics & Inventory Management', 'Management', 'Mentoring',
       'Negotiations & Efficiency', 'Organizing Events',
       'Pharmaceutical Activities', 'Photography', 'Planning & Execution',
       'Policy Making', 'Presentation Skills', 'Problem Solving',
       'Public Relations', 'Public Speaking', 'Sales',
       'Sign Language Expertise', 'Social Media Strategy & Marketing',
       'Sta

In [60]:
df2_encoded.head()

City    Rating  Arts and Crafts  Auditing  Beautician  \
0    69  2.618342                1         0           0   
1   116  2.168675                0         0           0   
2    69  2.811057                0         0           0   
3    69  2.554104                0         0           0   
4    69  2.811057                0         0           0   

   Branding & Marketing  Coaching  Community Outreach  Conflict Management  \
0                     0         0                   0                    0   
1                     0         0                   0                    0   
2                     0         0                   0                    0   
3                     1         0                   0                    0   
4                     0         0                   0                    0   

   Content Writing  ...  Fundraising and Donations  March  \
0                0  ...                          1      0   
1                0  ...                          1      0   
2                0  ...                          0      0   
3                0  ...                          0      0   
4                0  ...                          0      0   

   Milkar SDG's Champions  Rallies  Social Responsibility  Teaching  Vigils  \
0                       0        0                      0         1       0   
1                       0        0                      0         0       0   
2                       0        0                      1         0       0   
3                       0        0                      0         1       0   
4                       0        0                      0         0       0   

   Walks  Workshops  Organization  
0      0          1             0  
1      0          0             0  
2      0          0             0  
3      0          1             0  
4      0          0             0  

[5 rows x 90 columns]

In [61]:
print(df2_encoded.columns)

Index(['City', 'Rating', 'Arts and Crafts', 'Auditing', 'Beautician',
       'Branding & Marketing', 'Coaching', 'Community Outreach',
       'Conflict Management', 'Content Writing', 'Cooking/ Baking',
       'Counseling', 'Dancing', 'Data Analysis', 'Data Gathering',
       'Database Management', 'Digital Media Management',
       'Editing/proofreading', 'Event Management', 'Financial Skills',
       'Fundraising', 'Goal Setting', 'Graphic Designing',
       'Health/Medical Experience', 'Human Resource Skills',
       'Innovation & Creativity (New Ideas)', 'Interviewing', 'Leadership',
       'Logistics & Inventory Management', 'Management', 'Mentoring',
       'Negotiations & Efficiency', 'Organizing Events',
       'Pharmaceutical Activities', 'Photography', 'Planning & Execution',
       'Policy Making', 'Presentation Skills', 'Problem Solving',
       'Public Relations', 'Public Speaking', 'Sales',
       'Sign Language Expertise', 'Social Media Strategy & Marketing',
       'Sta

## Similarity Calculations

In [62]:
from sklearn.metrics.pairwise import cosine_similarity

In [63]:
# Compute cosine similarity
similarity_matrix = cosine_similarity(df2_encoded, df1_encoded)

In [64]:
# Convert similarity matrix to DataFrame
similarity_df = pd.DataFrame(similarity_matrix, index=df2_encoded.index, columns=df1_encoded.index)

In [65]:
similarity_df.head()

0         1         2         3         4         5         6    \
0  0.523114  0.410444  0.616004  0.523661  0.511805  0.538899  0.340439   
1  0.524229  0.411097  0.616985  0.524495  0.512854  0.539634  0.341165   
2  0.523477  0.410847  0.616609  0.524066  0.512236  0.539030  0.340747   
3  0.523247  0.410452  0.616016  0.523781  0.511826  0.538917  0.340455   
4  0.523586  0.410847  0.616609  0.524175  0.512274  0.539143  0.340747   

        7         8         9    ...       692       693       694       695  \
0  0.791685  0.523661  0.802427  ...  0.513398  0.526517  0.343140  0.817508   
1  0.792945  0.524495  0.803921  ...  0.514354  0.527639  0.343885  0.819396   
2  0.792296  0.524066  0.803016  ...  0.513882  0.526993  0.343376  0.817593   
3  0.791867  0.523781  0.802622  ...  0.513414  0.526541  0.343241  0.817572   
4  0.792462  0.524175  0.803184  ...  0.513882  0.526993  0.343448  0.817765   

        696       697       698       699       700       701  
0  0.687414  0.513398  0.988086  0.986881  0.987473  0.589976  
1  0.688907  0.514354  0.989244  0.988985  0.989311  0.591144  
2  0.687888  0.513774  0.988435  0.987567  0.988197  0.590114  
3  0.687616  0.513522  0.988106  0.987134  0.987713  0.590010  
4  0.688032  0.513882  0.988642  0.987774  0.988404  0.590238  

[5 rows x 702 columns]

In [66]:
type(similarity_df)

pandas.core.frame.DataFrame

In [67]:
similarity_df.shape

(7669, 702)

## Generate Recommendations

In [68]:
import numpy as np
from fuzzywuzzy import process

## Load Ground_Truth_Data.csv

In [69]:
# Load and process Ground_Truth_Data.csv
gt_df = pd.read_csv("Ground_Truth_Data_3.csv")
# Normalize event names and volunteer names
gt_df['Event_Name'] = gt_df['Event_Name'].str.strip().str.lower()
# Assuming Volunteers_Participated is stored as a string (e.g., "vol1, vol2, vol3")
def process_volunteers(x):
    return [v.strip().lower() for v in x.split(",")] if isinstance(x, str) else []
gt_df['Volunteers_Participated'] = gt_df['Volunteers_Participated'].apply(process_volunteers)

# Build dictionary: {event_name: list_of_volunteer_names}
ground_truth_dict = dict(zip(gt_df['Event_Name'], gt_df['Volunteers_Participated']))

## Recommendation Function

In [70]:
def get_volunteer_recommendations_2(df1, df1_encoded, df2, df2_encoded, similarity_df, ground_truth_dict=None, n_recommendations=5, weights=None):
    """
    Generates volunteer recommendations for a given event.
    
    Incorporates:
      - Content-based matching: location, skills/event types, rating.
      - Precomputed similarity scores.
      - A participation component (using normalized 'Activities_Volunteered' from df2).
      - If ground_truth_dict (a dict mapping event names to lists of volunteer names) is provided and contains the event,
        restricts the recommendation pool to that set of volunteers.
    
    Cases handled:
      - Case 1 (Few Volunteers): For very low participation counts, a moderate default is applied.
      - Case 2 (No Past Participation): The participation component will be near zero, so other features dominate.
      - Case 3 (Cold Start): New volunteers or events rely on content-based features and similarity scores.
      - Additionally, if an event has ground truth data for Volunteers_Participated, the recommendations are chosen only from that pool.
    """
    # Extend weights to include participation
    if weights is None:
        weights = {
            'location': 0.3,
            'skills_event_types': 0.3,
            'rating': 0.2,
            'similarity': 0.3,
            'participation': 0.3
        }
    try:
        # Get event name input from the console
        event_name = input("Enter Event Name: ").strip().lower()
        
        # Validate event existence in df1 (Volunteer_Activities.csv)
        if event_name not in df1['Title'].str.strip().str.lower().values:
            suggestions = process.extract(event_name, df1['Title'].str.lower().values, limit=3)
            print(f"Event '{event_name}' not found. Did you mean:")
            for suggestion in suggestions:
                print(f"- {suggestion[0]} (Confidence: {suggestion[1]}%)")
            return None
        
        # Get event index using normalized title comparison
        event_idx = df1[df1['Title'].str.strip().str.lower() == event_name].index[0]
        event_features = df1_encoded.loc[event_idx]
        
        # Restrict volunteer pool if ground_truth_dict is provided and the event has ground truth data
        if ground_truth_dict is not None and event_name in ground_truth_dict:
            pool_names = [name.strip().lower() for name in ground_truth_dict[event_name]]
            # Create a boolean mask for volunteers in df2 whose names appear in the ground truth pool
            mask = df2['Name'].str.strip().str.lower().isin(pool_names)
            if mask.sum() > 0:
                df2_pool = df2[mask].reset_index(drop=True)
                df2_encoded_pool = df2_encoded[mask].reset_index(drop=True)
                print(f"Using ground truth volunteer pool for event '{event_name}'. Total volunteers in pool: {len(df2_pool)}")
            else:
                print(f"Ground truth volunteer pool for event '{event_name}' is empty after filtering. Using full pool.")
                df2_pool = df2.copy()
                df2_encoded_pool = df2_encoded.copy()
        else:
            df2_pool = df2.copy()
            df2_encoded_pool = df2_encoded.copy()
        
        # Define feature columns (excluding non-informative ones)
        feature_cols = [col for col in df1_encoded.columns if col not in ['City', 'Rating', 'Organization']]
        
        # ----- Content-Based Components -----
        # 1. Location matching
        location_scores = (df2_encoded_pool['City'] == event_features['City']).astype(float) * weights['location']
        
        # 2. Skills and Event Types matching
        matching_features = (df2_encoded_pool[feature_cols].values * event_features[feature_cols].values).sum(axis=1)
        total_event_features = event_features[feature_cols].sum()
        feature_scores = (matching_features / total_event_features) * weights['skills_event_types'] if total_event_features > 0 else 0
        
        # 3. Rating contribution
        rating_scores = (df2_encoded_pool['Rating'] / 5.0) * weights['rating']
        
        # Combine content-based scores
        relevance_scores = location_scores + feature_scores + rating_scores
        
        # ----- Participation Component -----
        # Use "Activities_Volunteered" (count) from Volunteers_Data.csv (df2_pool) as participation
        participation_raw = df2_pool['Activities Volunteered'].astype(float)
        max_participation = participation_raw.max() if participation_raw.max() > 0 else 1
        participation_scores = (participation_raw / max_participation)
        
        # Case 1: For very low participation, assign a moderate default value to avoid penalizing too harshly.
        participation_scores = participation_scores.apply(
            lambda x: 0.5 if x < (1 / max_participation) else x
        )
        participation_scores = participation_scores * weights['participation']
        
        # ----- Similarity Component -----
        # Note: similarity_df is computed over all volunteers in the full pool.
        # We need to select rows corresponding to the volunteer pool (based on the filtering above).
        # If the original similarity_df has the same ordering as df2, then after filtering, we must reindex similarity scores.
        # One approach is to use df2_pool's index to get the corresponding similarity scores.
        # Here, we assume that df2 and similarity_df are aligned. So we filter similarity_df accordingly.
        similarity_scores_full = similarity_df.iloc[:, event_idx].values
        # Create an index mapping from full df2 to the pool:
        pool_indices = df2_pool.index.values
        similarity_scores = similarity_scores_full[pool_indices]
        
        # ----- Final Score Combination -----
        final_scores = weights['similarity'] * similarity_scores + relevance_scores + participation_scores
        
        # Get top N volunteers based on final scores.
        top_indices = np.argsort(final_scores)[-n_recommendations:][::-1]
        recommendations = pd.DataFrame({
            'Volunteer Name': df2_pool.iloc[top_indices]['Name'].values,
            'Total Score': final_scores[top_indices].round(3)
        })
        
        return recommendations.sort_values('Total Score', ascending=False)
    
    except Exception as e:
        print(f"An error occurred: {str(e)}")
        return None


## Get Recommendations

In [ ]:
if __name__ == "__main__":
    recommendations = get_volunteer_recommendations_2(
        df1=df1,
        df1_encoded=df1_encoded,
        df2=df2,
        df2_encoded=df2_encoded,
        similarity_df=similarity_df,
        ground_truth_dict = ground_truth_dict,
        n_recommendations=5
    )
    
    if recommendations is not None:
        print("\nRecommended Volunteers:")
        print(recommendations)


## Evaluation Metrics

In [83]:
import pandas as pd
import numpy as np
from fuzzywuzzy import process

def get_volunteer_recommendations_2(df1, df1_encoded, df2, df2_encoded, similarity_df, ground_truth_dict=None, n_recommendations=5, weights=None):
    """
    Generates volunteer recommendations for all events in ground_truth_dict.
    
    Incorporates:
      - Content-based matching: location, skills/event types, rating.
      - Precomputed similarity scores.
      - A participation component (using normalized 'Activities_Volunteered' from df2).
      - If ground_truth_dict (a dict mapping event names to lists of volunteer names) is provided and contains the event,
        restricts the recommendation pool to that set of volunteers.
    
    Cases handled:
      - Case 1 (Few Volunteers): For very low participation counts, a moderate default is applied.
      - Case 2 (No Past Participation): The participation component will be near zero, so other features dominate.
      - Case 3 (Cold Start): New volunteers or events rely on content-based features and similarity scores.
      - Additionally, if an event has ground truth data for Volunteers_Participated, the recommendations are chosen only from that pool.
    """
    # Extend weights to include participation
    if weights is None:
        weights = {
            'location': 0.3,
            'skills_event_types': 0.3,
            'rating': 0.2,
            'similarity': 0.3,
            'participation': 0.3
        }
    
    all_recommendations = []

    try:
        # Iterate over all events in ground_truth_dict
        for event_name in ground_truth_dict.keys():
            event_name = event_name.strip().lower()
            
            # Validate event existence in df1 (Volunteer_Activities.csv)
            if event_name not in df1['Title'].str.strip().str.lower().values:
                suggestions = process.extract(event_name, df1['Title'].str.lower().values, limit=3)
                print(f"Event '{event_name}' not found. Did you mean:")
                for suggestion in suggestions:
                    print(f"- {suggestion[0]} (Confidence: {suggestion[1]}%)")
                continue
            
            # Get event index using normalized title comparison
            event_idx = df1[df1['Title'].str.strip().str.lower() == event_name].index[0]
            event_features = df1_encoded.loc[event_idx]
            
            # Restrict volunteer pool if ground_truth_dict is provided and the event has ground truth data
            pool_names = [name.strip().lower() for name in ground_truth_dict[event_name]]
            mask = df2['Name'].str.strip().str.lower().isin(pool_names)
            df2_pool = df2[mask].reset_index(drop=True)
            df2_encoded_pool = df2_encoded[mask].reset_index(drop=True)

            if df2_pool.empty:
                print(f"Ground truth volunteer pool for event '{event_name}' is empty after filtering. Skipping this event.")
                continue
            
            print(f"Using ground truth volunteer pool for event '{event_name}'. Total volunteers in pool: {len(df2_pool)}")
            
            # Define feature columns (excluding non-informative ones)
            feature_cols = [col for col in df1_encoded.columns if col not in ['City', 'Rating', 'Organization']]
            
            # ----- Content-Based Components -----
            # 1. Location matching
            location_scores = (df2_encoded_pool['City'] == event_features['City']).astype(float) * weights['location']
            
            # 2. Skills and Event Types matching
            matching_features = (df2_encoded_pool[feature_cols].values * event_features[feature_cols].values).sum(axis=1)
            total_event_features = event_features[feature_cols].sum()
            feature_scores = (matching_features / total_event_features) * weights['skills_event_types'] if total_event_features > 0 else 0
            
            # 3. Rating contribution
            rating_scores = (df2_encoded_pool['Rating'] / 5.0) * weights['rating']
            
            # Combine content-based scores
            relevance_scores = location_scores + feature_scores + rating_scores
            
            # ----- Participation Component -----
            participation_raw = df2_pool['Activities Volunteered'].astype(float)
            max_participation = participation_raw.max() if participation_raw.max() > 0 else 1
            participation_scores = (participation_raw / max_participation)
            
            # Case 1: For very low participation, assign a moderate default value to avoid penalizing too harshly.
            participation_scores = participation_scores.apply(
                lambda x: 0.5 if x < (1 / max_participation) else x
            )
            participation_scores = participation_scores * weights['participation']
            
            # ----- Similarity Component -----
            similarity_scores_full = similarity_df.iloc[:, event_idx].values
            pool_indices = df2_pool.index.values
            
            # Ensure we only access valid indices
            if len(pool_indices) > 0:
                similarity_scores = similarity_scores_full[pool_indices]
            else:
                print(f"No valid indices found for similarity scores for event '{event_name}'. Skipping this event.")
                continue
            
            # ----- Final Score Combination -----
            final_scores = weights['similarity'] * similarity_scores + relevance_scores + participation_scores
            
            # Get top N volunteers based on final scores.
            top_indices = np.argsort(final_scores)[-n_recommendations:][::-1]
            recommendations = pd.DataFrame({
                'Volunteer Name': df2_pool.iloc[top_indices]['Name'].values,
                'Total Score': final_scores[top_indices].round(3),
                'Event Name': event_name
            })
            
            all_recommendations.append(recommendations)

        # Combine all recommendations into a single DataFrame
        final_recommendations = pd.concat(all_recommendations, ignore_index=True)
        final_recommendations.to_csv('Predicted_Recommendations.csv', index=False)
        print("Recommendations saved to 'Predicted_Recommendations.csv'.")

    except Exception as e:
        print(f"An error occurred: {str(e)}")

In [ ]:
if __name__ == "__main__":
    recommendations = get_volunteer_recommendations_2(
        df1=df1,
        df1_encoded=df1_encoded,
        df2=df2,
        df2_encoded=df2_encoded,
        similarity_df=similarity_df,
        ground_truth_dict = ground_truth_dict,
        n_recommendations=5
    )

In [90]:
import pandas as pd
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

# Load the predicted recommendations
predicted_df = pd.read_csv('Predicted_Recommendations.csv')

# Print the columns of the predicted DataFrame for debugging
print("Predicted DataFrame Columns:", predicted_df.columns)

# Load the ground truth data from Ground_Truth_Data.csv
ground_truth_df = pd.read_csv('Ground_Truth_Data_3.csv')

# Print the columns of the ground truth DataFrame for debugging
print("Ground Truth DataFrame Columns:", ground_truth_df.columns)

# Convert the ground truth DataFrame into a format suitable for merging
ground_truth_data = []
for index, row in ground_truth_df.iterrows():
    event = row['Event_Name']
    # Check if Volunteers_Participated is not NaN and is a string
    if pd.notna(row['Volunteers_Participated']):
        volunteers = str(row['Volunteers_Participated']).split(',')  # Convert to string and split
        for volunteer in volunteers:
            ground_truth_data.append({'Event': event.strip(), 'Volunteer': volunteer.strip(), 'Actual': 1})  # Actual = 1 for ground truth

# Create a DataFrame from the ground truth data
ground_truth_df = pd.DataFrame(ground_truth_data)

# Rename columns in predicted_df to match ground_truth_df for merging
predicted_df.rename(columns={
    'Event Name': 'Event',
    'Volunteer Name': 'Volunteer'
}, inplace=True)

# Merge predicted recommendations with ground truth data
merged_df = pd.merge(predicted_df, ground_truth_df, on=['Event', 'Volunteer'], how='left')

# Fill NaN values in 'Actual' column with 0 (indicating not in ground truth)
merged_df['Actual'] = merged_df['Actual'].fillna(0)

# Calculate predicted values based on whether the volunteer is in the ground truth
merged_df['Predicted'] = (merged_df['Total Score'] > 0).astype(int)

# Calculate evaluation metrics
y_true = merged_df['Actual'].values  # Ground truth values
y_pred = merged_df['Predicted'].values  # Predicted values

# Calculate evaluation metrics
accuracy = accuracy_score(y_true, y_pred)
precision = precision_score(y_true, y_pred)
recall = recall_score(y_true, y_pred)
f1 = f1_score(y_true, y_pred)
conf_matrix = confusion_matrix(y_true, y_pred)

# Print the results
print("Evaluation Metrics:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1 Score: {f1:.4f}")
print("Confusion Matrix:")
print(conf_matrix)

Predicted DataFrame Columns: Index(['Volunteer Name', 'Total Score', 'Event Name'], dtype='object')
Ground Truth DataFrame Columns: Index(['Event_Name', 'Volunteers_Participated'], dtype='object')
Evaluation Metrics:
Accuracy: 0.9553
Precision: 0.9553
Recall: 1.0000
F1 Score: 0.9772
Confusion Matrix:
[[   0   96]
 [   0 2054]]
